### ISSUE TITLE
Cannot find module '.../.next/server/static/U1ODQPfQegE-wf5TV0bgZ/pages/_document'

### ISSUE THREAD

**Issue by dancancro** [ooade/NextSimpleStarter#100]
I got this error right away after downloading the starter and running `npm run build` and `npm run start`.

```
{ Error: Cannot find module '/Users/Dan/work/b/original/.next/server/static/yOGv7a1uTpivdY1RQ64A1/pages/_document'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:582:15)
    at Function.Module._load (internal/modules/cjs/loader.js:508:25)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:22:18)
    at Object.loadComponents (/Users/Dan/work/b/original/node_modules/next-server/dist/server/load-components.js:16:24)
    at Server.renderToHTMLWithComponents (/Users/Dan/work/b/original/node_modules/next-server/dist/server/next-server.js:198:48)
    at Server.renderErrorToHTML (/Users/Dan/work/b/original/node_modules/next-server/dist/server/next-server.js:228:21)
    at Server.renderError (/Users/Dan/work/b/original/node_modules/next-server/dist/server/next-server.js:221:33)
    at Server.render404 (/Users/Dan/work/b/original/node_modules/next-server/dist/server/next-server.js:237:21)
    at Object.fn (/Users/Dan/work/b/original/node_modules/next-server/dist/server/next-server.js:120:32)
    at /Users/Dan/work/b/original/node_modules/next-server/dist/server/router.js:23:36
    at Server.run (/Users/Dan/work/b/original/node_modules/next-server/dist/server/next-server.js:157:23)
    at Server.handleRequest (/Users/Dan/work/b/original/node_modules/next-server/dist/server/next-server.js:80:21)
    at Server.emit (events.js:189:13)
    at parserOnIncoming (_http_server.js:676:12)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:109:17) code: 'MODULE_NOT_FOUND' }
```

**Comment by dancancro** [comment]
`npm run dev` doesn't cause any errors. Both the original starter app and the one to which I added my code start up and run fine, but there is no service worker registered or running in either.

**Comment by ooade** [comment]
My guess is you switch from `serverless` to `server` in your next config to run it locally. So, you'd need to clear your `.next` directory to make this work.

Thanks, and I'll try to clear thing up for easy usage.

**Comment by dancancro** [comment]
That overcame that problem but now when I run it, I get this in the browser console:

```
A bad HTTP response code (404) was received when fetching the script.
main-a21604b48a41d27a172b.js:1 SW registration failed:  TypeError: Failed to register a ServiceWorker: A bad HTTP response code (404) was received when fetching the script.
```

I see a new service worker listed on the Application dev tools tab but it has a red 'x'. I got these errors without doing anything to the starter besides your recommendation. So you ought to be able to reproduce the problem.

Also if I'm understanding these words correctly, I think I need something "serverless" because I deploy the app on Netlify. What are the effects of this setting? I have been using `next export` for that and got the same error message when trying that with this starter.

Service workers seem to be [supported on Netlify](https://www.netlify.com/blog/2017/10/31/service-workers-explained/).

**Comment by ooade** [comment]
Yeah, it wouldn't work directly because of the service worker path. 

That's why we have this in [now.json](https://github.com/ooade/NextSimpleStarter/blob/master/now.json)

```json
{
 "src": "^/service-worker.js$",
 "dest": "/_next/static/service-worker.js",
 "headers": {
 "cache-control": "public, max-age=43200, immutable",
 "Service-Worker-Allowed": "/"
 }
}
```

So you'd have map the service worker source to where it's destination. I wouldn't mind a PR to make next-offline generate our service worker in static directory so we wouldn't be dealing with these issues. Initially I had an express server to do that (i.e. get the path, and resolve). But there's really no need for a server that's why I took it off.

**Comment by dancancro** [comment]
Do you know how to solve my problem? I don't understand clearly what "map the service worker source to where it's destination" means. I have not changed anything in your code. So you may give me exact, literal instructions.

**Comment by ooade** [comment]
Here is how to go about solving it:

- Go to next.config.js, change `serverless` to `server`.
- Run build script `npm run build`

**Comment by ooade** [comment]
Note that the service worker will be generated at `/static/service-worker.js`

**Comment by dancancro** [comment]
It's not working for me. Is it working for you? From the top project directory I run `npm run build && cd out && serve -p 3000`. I see a file `./out/static/service-worker.js` has been generated. I open `localhost:3000` in the browser. It gives me this error message:
```
A bad HTTP response code (404) was received when fetching the script.
main-a21604b48a41d27a172b.js:1 SW registration failed:  TypeError: Failed to register a ServiceWorker: A bad HTTP response code (404) was received when fetching the script.
```

I have changed nothing about your starter besides changing `serverless` to `server` on line 4 of `next.config.js`.

**Comment by ooade** [comment]
I would have to document how to deploy to several hosting services. But for now use [this](https://www.netlify.com/blog/2019/01/16/redirect-rules-for-all-how-to-configure-redirects-for-your-static-site/) to set a redirect rule from `/static/service-worker.js` to `/service-worker.js`.

**Comment by ooade** [comment]
I would have to document how to deploy to several hosting services. But for now use [this](https://www.netlify.com/blog/2019/01/16/redirect-rules-for-all-how-to-configure-redirects-for-your-static-site/) to guide you to setting a redirect rule from `/static/service-worker.js` to `/service-worker.js`.

**Comment by dancancro** [comment]
I get this problem running on localhost. Do you get this problem running on localhost?

**Comment by ooade** [comment]
Unfortunately, to run this on localhost with the current setup is nearly impossible. Mainly because we went serverless. 

Two ways I know to go about it locally
1. Install and use [superstatic](https://www.npmjs.com/package/superstatic) to serve and redirect for you
2. Install and use express to resolve `/static/service-worker.js` to `/service-worker.js`.

In ~3 hours, I'll add the first one so we could all test locally, as well.

**Comment by ooade** [comment]
Hi @dancancro! I've updated the repo.

Just run the commands:
```sh
npm run export
npm run start:static
```

And you're good to go. Thanks!

**Comment by ooade** [comment]
I'll update the docs or you could help to add the instructions. And perhaps, a little guide to uploading to Netlify. Same commands.

**Comment by ooade** [comment]
Oops! I added the guide already https://github.com/ooade/NextSimpleStarter#deploy-to-netlify

Cheers 🎉 🎉

**Comment by dancancro** [comment]
@ooade Thanks! [It works](https://www.stopindoctrination.org)

I'm not seeing a big performance improvement yet. The Chrome Lighthouse score might have gone up a little, and the service worker is running. The main problem has been and continues to be loading the app, specifically on my phone and from the version that opens using an app icon rather than from Safari. I tried deleting that icon and saving it anew but then I got an app icon with a screen shot of the page instead of a favicon, and touching it just opens a new page of safari along with the browser controls. That's a separate matter, but anyway, thanks again.
